File: libhose/Worker.cpp
#include <string.h>
#include <zmq.hpp>
#include <iostream>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/wait.h>
#include <map>
#include <sstream>
#include <vector>
#include "Worker.h"
#include "util.h"


void callSocket(::std::string* request, ::std::string* response, ::zmq::socket_t* socket, int timeout) {
    ::std::cout << "" << " Sending " << *request << ::std::endl;

    // setting up the poller for this exchange

    zmq_pollitem_t poll_items[1];
    zmq_pollitem_t item1;
    item1.socket = *socket;
    item1.events = ZMQ_POLLIN;
    poll_items[0] = item1;

    // sending the message to the master
    ::zmq::message_t msg;
    str2msg(request, &msg);
    try {
        socket->send(msg, ZMQ_NOBLOCK);
    }
    catch (...) {
        ::std::cout << "" << " could not send a msg" << ::std::endl;
        throw RegisterError();
    }

    ::std::cout << "" << " waiting for an answer" << ::std::endl;

    // waiting for an answer now
    ::zmq::poll(poll_items, 1, timeout);

    ::std::cout << "" <<  " poll is over" << ::std::endl;
    // no answer in time
    if (poll_items[0].revents == 0) {
        ::std::cout << "did not get anything" << ::std::endl;
        throw RegisterError();
    }
    else {
    // getting the result
    ::zmq::message_t res;
    socket->recv(&res);

    ::std::string result;
    msg2str(&res, &result);

    ::std::cout <<"" <<  " we got " << result << ::std::endl;
    ::std::cout << "" << " we want " << *response << ::std::endl;

    if (result != *response) {
        throw RegisterError();
    }
    else ::std::cout << "" << " Registered!" << ::std::endl;
    }
}

/*
    heartbeat -- pings the master 
*/
void *heartbeat(void *ptr) {
Worker *worker = (Worker*) ptr;
::std::string pong = "PONG";
::std::vector< ::std::string> vreq;
vreq.push_back("PING");
vreq.push_back(worker->receiverChannel);
::std::string req;
serialize(&vreq, &req);
int failures = 0;
int max_failures = 10;

while (worker->heartbeatRunning && failures < max_failures) {
    try {
        callSocket(&req, &pong, worker->endpoint, worker->timeout);
        ::std::cout << "ping did work" << ::std::endl;
    }
    catch (...) {
        ::std::cout << "ping did not work!" << ::std::endl;
        failures += 1;
    }
    // we need to sleep here
    sleep(worker->heartbeatDelay);
}
// if we quit because of failures, we need to toggle the worker flag
if (failures >= max_failures) {
    worker->heartbeatFailed = true;
}
::std::cout << "bye!" << ::std::endl;
}


/*
* Workers Class
*
*/

Worker::Worker(const char* receiverChannel, const char* endPoint) {
this->heartbeatDelay = 10;
this->timeout = 1000000;  // zmq timeout is in microseconds
this->ctx = new ::zmq::context_t(1);
this->receiver = new ::zmq::socket_t(*this->ctx, ZMQ_REP);
this->receiver->bind(receiverChannel);
this->receiverChannel = receiverChannel;
this->endpoint = new ::zmq::socket_t(*this->ctx, ZMQ_REQ);
this->endpoint->connect(endPoint);
this->heartbeatFailed = false;
zmq_pollitem_t item1;
item1.socket = *this->receiver;
item1.events = ZMQ_POLLIN;
this->poll_items[0] = item1;
}

Worker::~Worker() {
this->receiver->close();
delete this->receiver;
this->endpoint->close();
delete this->endpoint;
delete this->ctx;
}

void Worker::execute(::std::vector< ::std::string>* vreq,  ::std::vector< ::std::string>* vres) {
    vres->push_back("NOTIMPLEMENTED");
}

void Worker::reg() {
    // register to the master
    ::std::cout << "registering" << ::std::endl;
    ::std::vector< ::std::string> vreq;
    vreq.push_back("PING");
    vreq.push_back(this->receiverChannel);
    ::std::string req;
    serialize(&vreq, &req);
    ::std::string resp = "PONG";
    callSocket(&req, &resp, this->endpoint, this->timeout);
}

void Worker::run() {
// register
this->reg();

// wait a bit
sleep(1);

// start the heartbeat
pthread_t th;
::std::cout << "creating the HB" << ::std::endl;
this->heartbeatRunning = true;
pthread_create(&th, NULL, heartbeat, (void*)this);

::std::cout << "Now waiting for some job" << ::std::endl;
// now loop and wait for some work to do
zmq_pollitem_t poll_items[1];
zmq_pollitem_t item1;
item1.socket = *this->receiver;
item1.events = ZMQ_POLLIN;
poll_items[0] = item1;

::std::vector< ::std::string> vresp;
::std::vector< ::std::string> vreq;

this->running = true;

while (this->running && !this->heartbeatFailed) {
    // waiting for an answer now
    ::zmq::poll(poll_items, 1, this->timeout);

    for (short j = 0; j < poll_items[0].revents; j++) {

    vresp.clear();
    vreq.clear();

    recv(this->receiver, &vreq);

    if (vreq.at(0) != "JOB") {
        // error, raise something
    }

    vresp.push_back("JOBRES");

    try {
        this->execute(&vreq, &vresp);
    }
    catch (...) {
        vresp.push_back("ERROR");
    }
    // sending back the response
    send(this->receiver, &vresp);
    }
}

// stop the heartbeat
::std::cout << "stopping the heartbeat" << ::std::endl;
this->heartbeatRunning = false;
pthread_join(th, NULL);
::std::cout << "The worker is done" << ::std::endl;
}

File: powerhose/client/worker.py
import time
import threading

from gevent_zeromq import zmq

from powerhose.client.pinger import Pinger
from powerhose.util import serialize, unserialize
from powerhose import logger


class RegisterError(Exception):
    pass


class Worker(object):

    def __init__(self, endpoint, identity, target, timeout=1.):
        self.identity = identity
        self.ctx = zmq.Context()
        self.timeout = timeout * 1000
        self.master = self.ctx.socket(zmq.REQ)
        self.master.connect(endpoint)
        self.work = self.ctx.socket(zmq.REP)
        self.work.bind(identity)
        self.registered = self.running = False
        # setting a poller
        self.poller = zmq.Poller()
        self.poller.register(self.work, zmq.POLLIN)
        self.locker = threading.RLock()
        self.pinger = Pinger(self.identity, self.master, self.locker,
                             self.failed)
        self.target = target

    def failed(self):
        logger.debug("ping failed let's die")
        try:
            self._msg('REMOVE', 'REMOVED')
        except RegisterError:
            pass
        self.stop()

    def register(self):
        self._msg('PING', 'PONG')
        self.registered = True

    def _msg(self, req, rep):
        self.pinger.disable()
        try:
            with self.locker:
                poller = zmq.Poller()
                poller.register(self.master, zmq.POLLIN)

                # ping the master we are online, with an ID
                try:
                    data = serialize(req, self.identity)
                    self.master.send(data, zmq.NOBLOCK)
                except zmq.ZMQError, e:
                    raise RegisterError(str(e))

                try:
                    events = dict(poller.poll(self.timeout))
                except zmq.ZMQError, e:
                    raise RegisterError(str(e))

                if events == {}:
                    raise RegisterError("Nothing came back")
                else:
                    for socket in events:
                        res = socket.recv()
                        if res != rep:
                            raise RegisterError(res)
        finally:
            self.pinger.enable()

    def stop(self):
        self.running = False
        self.pinger.stop()
        time.sleep(.1)
        self.ctx.destroy(0)

    def run(self):
        self.running = True
        self.register()
        self.pinger.start()

        while self.running and not self.pinger.unresponsive:
            try:
                events = dict(self.poller.poll(self.timeout))
            except zmq.ZMQError:
                break

            for socket in events:
                msg = unserialize(socket.recv())

                logger.debug(msg)
                if msg[0] == 'JOB':
                    # do the job and send the result
                    start = time.time()
                    try:
                        res = self.target(msg[1:])
                    except Exception, e:
                        # XXX log the error
                        res = str(e)
                    logger.debug('%.6f' % (time.time() - start))
                    socket.send(serialize("JOBRES", msg[1], res))
                else:
                    socket.send('ERROR')

        try:
            self._msg('REMOVE', 'REMOVED')
        except RegisterError:
            pass
